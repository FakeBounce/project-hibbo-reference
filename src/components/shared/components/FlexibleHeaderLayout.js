import React, { PureComponent } from 'react';
import { ScrollView, Animated, StyleSheet } from 'react-native';
import {
  func,
  node,
  string,
  number,
  oneOfType,
  object,
  array,
} from 'prop-types';

import theme from 'styles/appStyles';
import { StyledContainerBasic } from 'styles/styledComponents/containers';
import {
  AnimatedFlexibleHeaderView,
  SFlexHeaderMain,
  SFlexBackArrow,
  SFlexHeaderTitle,
  SFlexTitle,
  SFlexHeaderTitleRight,
} from '../styles';
import Back from '../Back';

const styles = StyleSheet.create({
  scrollViewContentContainer: {
    flexGrow: 1,
  },
});

class FlexibleHeaderLayout extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      headerHeight: new Animated.Value(this.props.maxHeight),
    };

    this.headerAnimationOngoing = false;
    this.headerClosed = false;
  }

  componentWillReceiveProps(nextProps) {
    const { scrollPosY } = nextProps;
    if (!this.headerAnimationOngoing) {
      if (scrollPosY > 10 && !this.headerClosed) {
        this.headerAnimationOngoing = true;
        Animated.timing(this.state.headerHeight, {
          toValue: this.props.minHeight,
          duration: 150,
        }).start(() => {
          this.headerAnimationOngoing = false;
          this.headerClosed = true;
        });
      }

      if (scrollPosY <= 10 && this.headerClosed) {
        this.headerAnimationOngoing = true;
        Animated.timing(this.state.headerHeight, {
          toValue: this.props.maxHeight,
          duration: 150,
        }).start(() => {
          this.headerAnimationOngoing = false;
          this.headerClosed = false;
        });
      }
    }
  }

  render() {
    return (
      <StyledContainerBasic style={this.props.style}>
        <AnimatedFlexibleHeaderView
          style={[
            {
              backgroundColor: this.state.headerHeight.interpolate({
                inputRange: [this.props.minHeight, this.props.maxHeight],
                outputRange: [theme.colors.moneyPotText, theme.colors.white],
              }),
              height: this.state.headerHeight,
            },
          ]}
        >
          <SFlexHeaderMain>
            <SFlexBackArrow>
              <Back
                type={this.props.arrowType}
                action={this.props.onBack}
                color={this.props.arrowColor}
              />
            </SFlexBackArrow>
            <SFlexHeaderTitle>
              <SFlexTitle>{this.props.title}</SFlexTitle>
            </SFlexHeaderTitle>
            <SFlexHeaderTitleRight />
          </SFlexHeaderMain>
          {this.props.menuChildren}
        </AnimatedFlexibleHeaderView>

        <ScrollView
          showsVerticalScrollIndicator={false}
          onScroll={this.props.onScroll}
          scrollEventThrottle={14}
          contentContainerStyle={styles.scrollViewContentContainer}
        >
          {this.props.children}
        </ScrollView>
      </StyledContainerBasic>
    );
  }
}

FlexibleHeaderLayout.propTypes = {
  onBack: func.isRequired,
  onScroll: func.isRequired,
  children: node.isRequired,
  menuChildren: node,
  title: string.isRequired,
  style: oneOfType([object, number, array]),
  minHeight: number,
  maxHeight: number,
  scrollPosY: number,
  arrowColor: string,
  arrowType: string,
};

FlexibleHeaderLayout.defaultProps = {
  style: {},
  menuChildren: null,
  minHeight: 60,
  maxHeight: 120,
  scrollPosY: 0,
  arrowColor: 'black',
  arrowType: 'straight',
};

export default FlexibleHeaderLayout;
