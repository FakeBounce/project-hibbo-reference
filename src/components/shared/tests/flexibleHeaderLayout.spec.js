import { View, Text } from 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import FlexibleHeaderLayout from '../components/FlexibleHeaderLayout';

describe('FlexibleHeaderLayout component', () => {
  it('should render the component', () => {
    const fn = jest.fn();

    const value = renderer
      .create(
        <FlexibleHeaderLayout onBack={fn}>
          <View>
            <Text>Test</Text>
          </View>
        </FlexibleHeaderLayout>,
      )
      .toJSON();

    expect(value).toMatchSnapshot();
  });
});
