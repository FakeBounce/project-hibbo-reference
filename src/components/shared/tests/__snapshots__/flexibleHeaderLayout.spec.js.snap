// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexibleHeaderLayout component should render the component 1`] = `
<View
  style={
    Array [
      undefined,
      Object {},
    ]
  }
>
  <View
    collapsable={undefined}
    style={
      Object {
        "backgroundColor": "rgba(255, 255, 255, 1)",
        "height": 90,
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Image
        source={1}
      />
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={undefined}
    />
  </View>
  <RCTScrollView
    contentContainerStyle={undefined}
    onScroll={[Function]}
    scrollEventThrottle={16}
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
        >
          Test
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
